PXB-2681 PXB gets stuck during incremental backup prepare of encrypte…

rahulmalik87 opened the pull request:
…d tables

https://jira.percona.com/browse/PXB-2681

Problem:
With pagetracking, PXB crashes during prepare of incremental backup if alter
encryption tablespace='y' happens after the checkpoint.

Analysis:
If modifed pages due to change in encryption are not checkpoint, then the list
of those pages would not be available with the page tracking, and it relies on
redo to apply those changes to the tablespace. The redo apply on page 0 reset
the in-memory encryption, and because of that, remaining pages could not be
decrypted as it doesn't have encryption information.

Fix:
ALTER tablespace encryption='y' DDL forcefully flushes all the pages to disk at
the end. If this alter happens after the checkpoint, PXB scans complete ibd to
know the modified pages. So it doesn't need to apply redo on page 0 as it has
copied the page

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1830,6 +1830,16 @@ static byte *recv_parse_or_apply_log_rec_body(
             /* For cloned db header page has the encryption information. */
             !recv_sys->is_cloned_db) {
+          /* If Encryption info is changed after the last checkpoint lSN, scan
+          the tablespace to know modified pages. This scan and copy of modified
+          pages would avoid the apply the redo logs due to changed in encryption
+          info. As there is problem of apply redo on page 0 because it reset
+          in-memory encryption and page can't decrypted */
```
The point about scan and copying modified pages seems like an optimization. I think it is irrelevant to the current bug. I think this bug is because a full backup can have full encrypted tablespace and for the next incremental, if ALTER TABLESPACE ENCRYPTION='N' happens and if checkpoint doesn't move since fullbackup, it is possible that we apply page 0 encryption='n' change first. ie reset of encryption info first.

Then after this page 0 redo apply, we cannot read other encrypted pages from disk and cannot make them unencrypted.

satya-bodapati replied:
@rahulmalik87 ^ please address this review comment

rahulmalik87 replied:
fixed

satya-bodapati reviewed (changes requested):
Please paste this test result without and with patch. Also jenkins to see if other tests index_load redo tests are passing (full pxb suite).

rahulmalik87 commented:
> Please paste this test result without and with patch. Also jenkins to see if other tests index_load redo tests are passing (full pxb suite).

This is ASAN build
Running against Percona Server 8.0.28-19-debug (XtraDB 8.0.28-19)
Using 'xtrabackup' as xtrabackup binary
Using 8 parallel workers

==============================================================================
TEST                                   WORKER    RESULT     TIME(s) or COMMENT
------------------------------------------------------------------------------
encryption_after_checkpoint              w1	[failed]    38
==============================================================================
Spent 38 of 63 seconds executing testcases

SUMMARY: 1 run, 0 successful, 0 skipped, 1 failed

Failed tests:  encryption_after_checkpoint

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1830,6 +1830,16 @@ static byte *recv_parse_or_apply_log_rec_body(
             !fsp_is_system_or_temp_tablespace(space_id) &&
             /* For cloned db header page has the encryption information. */
             !recv_sys->is_cloned_db) {
+          /* If Encryption info is changed after the last checkpoint lSN, scan
+          the tablespace to know modified pages. This scan and copy of modified
+          pages would avoid the apply the redo logs due to changed in encryption
+          info. As there is problem of apply redo on page 0 because it reset
+          in-memory encryption and page can't decrypted */
```
```suggestion
          /* If tablespace has been in-place encrypted ie ALTER TABLESPACE
          ENCRYPTION=Y/N after the last checkpoint LSN (after full backup),
          and if the checkpoint LSN doesn't move between full and incremental
          backups, page tracking will not give the full list of modified pages. So
          we rely on the redo to roll forward the changes.
          
          With redo based apply, if page 0 redo is applied first, it will remove
          the encryption info and after this, other encrypted pages in tablespace
          cannot be decrypted.
          
          Scan and copy modified pages of such tablespaces during the incremental
          backup. This way we would have the latest state of tablespace and redo apply
          will skip all the redo generated on the tablespace. */
```

satya-bodapati commented:
@rahulmalik87 please also add Jenkins result

altmannmarcelo commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1830,6 +1830,16 @@ static byte *recv_parse_or_apply_log_rec_body(
             !fsp_is_system_or_temp_tablespace(space_id) &&
             /* For cloned db header page has the encryption information. */
             !recv_sys->is_cloned_db) {
+          /* If Encryption info is changed after the last checkpoint lSN, scan
```
Please add this within `#ifdef XTRABACKUP` block

altmannmarcelo commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1830,6 +1830,16 @@ static byte *recv_parse_or_apply_log_rec_body(
+          info. As there is problem of apply redo on page 0 because it reset
+          in-memory encryption and page can't decrypted */
+          if (opt_page_tracking && xtrabackup_incremental != nullptr &&
+              recv_sys->recovered_lsn > incremental_start_checkpoint_lsn) {
```
shouldn't we be checking here if we have encryption information in memory for this space_id in order to switch to full scan to avoid false positives here?

rahulmalik87 replied:
It is not required.  We switch to full scan if we find encryption redo on page 0 which has LSN more than checkpoint LSN, because those page would not track by page-tracking, but these pages are flushed to disk( As alter Encryption forcefully flushed before completing) .

altmannmarcelo replied:
what is not clear to me here is how you know that page0 changed from encrypted to non encrypted by just checking incremental_start_checkpoint_lsn and recovered_lsn?

rahulmalik87 replied:
in both case, we would switch to a full scan.

satya-bodapati replied:
@altmannmarcelo @rahulmalik87 As of now, mlog_write_string on page 0 is considered as encryption info. Upstream also treats this way. To be more accurate, another way is to determine the offset at which the mlog_write_string happens.

altmannmarcelo commented on storage/innobase/xtrabackup/test/suites/pagetracking/encryption_after_checkpoint.sh:
```diff
@@ -0,0 +1,61 @@
+mysql test <<EOF
+CREATE TABLE t1 (i serial) KEY_BLOCK_SIZE=2 ENCRYPTION='Y';
+CREATE TABLESPACE tab02k_e ADD DATAFILE 'tab02k_e.ibd'  FILE_BLOCK_SIZE 02k ENCRYPTION='Y';
+CREATE TABLE t2 (i serial) KEY_BLOCK_SIZE=2 TABLESPACE tab02k_e ENCRYPTION='Y';
```
Is this bug only reproducible with KEY_BLOCK_SIZE set?

rahulmalik87 replied:
No, corruption is seen without KEY_BLOCK_SIZE, but with KEY_BLOCK_SIZE is easily crashes also.

rahulmalik87 commented:
jenkin results https://pxb.cd.percona.com/job/percona-xtrabackup-8.0-test-param/197. I added the patch for check_for_passwords.sh. I will remove that while rebase.

satya-bodapati commented:
@rahulmalik87 pagetracking suite is not executed. Please execute this suite and other excluded suites. Also I think Jenkins cannot run against debug server. So please execute this test manually. We should also have a jenkins run that runs against debug server.

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1830,6 +1830,27 @@ static byte *recv_parse_or_apply_log_rec_body(
             !fsp_is_system_or_temp_tablespace(space_id) &&
             /* For cloned db header page has the encryption information. */
             !recv_sys->is_cloned_db) {
+#ifdef XTRABACKUP
+          /* If tablespace has been in-place encrypted ie ALTER TABLESPACE
+          ENCRYPTION=Y/N after the last checkpoint LSN (after full backup),
+          and if the checkpoint LSN doesn't move between full and incremental
+          backups, page tracking will not give the full list of modified pages.
+          So we rely on the redo to roll forward the changes.
+
+          With redo based apply, if page 0 redo is applied first, it will remove
+          the encryption info and after this, other encrypted pages in
+          tablespace cannot be decrypted.
+
+          Scan and copy modified pages of such tablespaces during the
+          incremental backup. This way we would have the latest state of
+          tablespace and redo apply will skip all the redo generated on the
+          tablespace. */
+          if (opt_page_tracking && xtrabackup_incremental != nullptr &&
+              recv_sys->recovered_lsn > incremental_start_checkpoint_lsn) {
```
aligment of this block looks wrong. PLease check @rahulmalik87

rahulmalik87 commented:
> @rahulmalik87 pagetracking suite is not executed. Please execute this suite and other excluded suites. Also I think Jenkins cannot run against debug server. So please execute this test manually. We should also have a jenkins run that runs against debug server.

this pagetracking suite is executed.  yeahThis test required debug build.  Will create jira ticket for tracking to run against debug server. 
https://pxb.cd.percona.com/job/percona-xtrabackup-8.0-test-param/197/CMAKE_BUILD_TYPE=RelWithDebInfo,DOCKER_OS=asan,XTRABACKUP_TARGET=innodb80/testReport/(root)/

satya-bodapati commented:
> > @rahulmalik87 pagetracking suite is not executed. Please execute this suite and other excluded suites. Also I think Jenkins cannot run against debug server. So please execute this test manually. We should also have a jenkins run that runs against debug server.
> 
> this pagetracking suite is executed. yeahThis test required debug build. Will create jira ticket for tracking to run against debug server. https://pxb.cd.percona.com/job/percona-xtrabackup-8.0-test-param/197/CMAKE_BUILD_TYPE=RelWithDebInfo,DOCKER_OS=asan,XTRABACKUP_TARGET=innodb80/testReport/(root)/

@rahulmalik87 for now, please execute against debug server locally and paste the result here

rahulmalik87 commented:
> > > @rahulmalik87 pagetracking suite is not executed. Please execute this suite and other excluded suites. Also I think Jenkins cannot run against debug server. So please execute this test manually. We should also have a jenkins run that runs against debug server.
> > 
> > 
> > this pagetracking suite is executed. yeahThis test required debug build. Will create jira ticket for tracking to run against debug server. https://pxb.cd.percona.com/job/percona-xtrabackup-8.0-test-param/197/CMAKE_BUILD_TYPE=RelWithDebInfo,DOCKER_OS=asan,XTRABACKUP_TARGET=innodb80/testReport/(root)/
> 
> @rahulmalik87 for now, please execute against debug server locally and paste the result here

Running against Percona Server 8.0.28-19-debug (XtraDB 8.0.28-19)
Using 'xtrabackup' as xtrabackup binary
Using 8 parallel workers

==============================================================================
TEST                                   WORKER    RESULT     TIME(s) or COMMENT
------------------------------------------------------------------------------
encryption_after_checkpoint              w1	[passed]    46
==============================================================================
Spent 46 of 66 seconds executing testcases

SUMMARY: 1 run, 1 successful, 0 skipped, 0 failed

See results/ for detailed output

satya-bodapati commented:
> > > > @rahulmalik87 pagetracking suite is not executed. Please execute this suite and other excluded suites. Also I think Jenkins cannot run against debug server. So please execute this test manually. We should also have a jenkins run that runs against debug server.
> > > 
> > > 
> > > this pagetracking suite is executed. yeahThis test required debug build. Will create jira ticket for tracking to run against debug server. https://pxb.cd.percona.com/job/percona-xtrabackup-8.0-test-param/197/CMAKE_BUILD_TYPE=RelWithDebInfo,DOCKER_OS=asan,XTRABACKUP_TARGET=innodb80/testReport/(root)/
> > 
> > 
> > @rahulmalik87 for now, please execute against debug server locally and paste the result here
> 
> Running against Percona Server 8.0.28-19-debug (XtraDB 8.0.28-19) Using 'xtrabackup' as xtrabackup binary Using 8 parallel workers
> ## ==============================================================================
> TEST WORKER RESULT TIME(s) or COMMENT
> # encryption_after_checkpoint w1 [passed] 46
> 
> Spent 46 of 66 seconds executing testcases
> 
> SUMMARY: 1 run, 1 successful, 0 skipped, 0 failed
> 
> See results/ for detailed output

@rahulmalik87 Thanks for pasting the result. Please create JEN issue for running against debug builds of PS/MySQL Server

satya-bodapati reviewed (approved):
@rahulmalik87 please update PR against the latest 8.0 branch or against the release branch. and then merge it.

Commit:
PXB-2681 PXB gets stuck during incremental backup prepare of encrypted tables
https://jira.percona.com/browse/PXB-2681

Problem:
With pagetracking, PXB crashes during prepare of incremental backup if alter
encryption tablespace='y' happens after the checkpoint.

Analysis:
If modifed pages due to change in encryption are not checkpoint, then the list
of those pages would not be available with the page tracking, and it relies on
redo to apply those changes to the tablespace. The redo apply on page 0 reset
the in-memory encryption, and because of that remaining pages could not be
decrypted as it doesn't have encryption information.

Fix:
ALTER tablespace encryption='y' DDL forcefully flushes all the pages to disk at
the end. If this alter happens after the checkpoint, PXB scans complete ibd to
know the modified pages. So it doesn't need to apply redo on page 0 as it has
copied the page

rahulmalik87 commented:
> @rahulmalik87 please update PR against the latest 8.0 branch or against the release branch. and then merge it.

Done @satya-bodapati, Going to merge to 8.0 branch
